File: stack/1024_remove_adj_dupes.py
class Solution(object):
    # bad solution, strings are immutable in python hence result += c
    # and result = result[:-1] create new strings which involve copying
    # all characters.
    # overall time complexity becomes O(n^2) (for loop + string operations)
    def removeDuplicates(self, s: str) -> str:
        result: str = ""
        for c in s:
            if result == "":
                result += c
                continue

            if result[-1] == c:
                result = result[:-1]
            else:
                result += c

        return result

    # use list as stack
    def removeDuplicates2(self, s: str) -> str:
        stack: list[str] = []
        for c in s:
            if stack and stack[-1] == c:
                stack.pop()
            else:
                stack.append(c)

        return "".join(stack)


sol = Solution()
print(Solution.removeDuplicates(sol, "abbaca"))

File: stack/1024_remove_adj_dupes.cpp
// 1024. Remove All Adjacent Duplicates In String

#include <algorithm>
#include <iostream>
#include <stack>
using namespace std;

class Solution {
   public:
    string removeDuplicates(string s) {
        stack<char> st;
        string result;

        for (char c : s) {
            if (!st.empty() && st.top() == c) {
                st.pop();
            } else {
                st.push(c);
            }
        }

        while (!st.empty()) {
            result += st.top();
            st.pop();
        }

        reverse(result.begin(), result.end());
        return result;
    }

    // better (use a string itself like a stack)
    string removeDuplicates2(string s) {
        string result;
        for (char c : s) {
            if (!result.empty() && result.back() == c) {
                result.pop_back();
            } else {
                result.push_back(c);
            }
        }

        return result;
    }
};

int main() {
    Solution sol;
    string s = "abbaca";
    cout << sol.removeDuplicates2(s) << endl;  // expected: ca

    return 0;
}